Add metadata considerations #63 (#78)
ietf-wg-privacypass/base-drafts: draft-ietf-privacypass-protocol.md

## Server setup {#server-setup}

Before the protocol takes place, the server chooses a ciphersuite and
generates a keypair by running `(pkS, skS) = KeyGen()`.

> The server and client also determine the public metadata that they will
> like to add, and the server is aware of the client metadata ahead of time.
> Metadata must be of the type `Metadata`, an opaque byte string of
> arbitrary length representing. This configuration must be available to all

Reviewer: ```suggestion
The client and server also agree on public metadata values used during the protocol.
See {{client-setup}} for more information about this metadata. This configuration must be available to all
```

clients that interact with the server (for the purpose of engaging in a
Privacy Pass exchange). We assume that the server has a public (and unique)
identity that the client uses to retrieve this configuration.

> In {{draft-davidson-pp-architecture}} we discuss the kind of public
> metadata that is allowed.

Reviewer: ```suggestion
```

Author: Done ;)

## Client setup {#client-setup}

The client initialises a global storage system `store` that allows it
store the tokens that are received during issuance. The storage system
is a map of server identifiers (`server.id`) to arrays of stored tokens.
We assume that the client knows the server public key `pkS` ahead of

> time. The server and client also determine the public metadata that they will
> like to add, and the client is aware of the server metadata ahead of time.
> Metadata must be of the type `Metadata`, an opaque byte string of
> arbitrary length representing. The client picks a value `m` of tokens to

Reviewer: ```suggestion
As above, the client and server also agree on public metadata values used during the protocol.
Each metadata value is an arbitrary-length byte string. The client picks an integer value `m > 0` of tokens to
```

receive during the issuance phase. In {{draft-davidson-pp-architecture}} we
discuss mechanisms that the client can use to ensure that this public key

> is consistent across the entire ecosystem, and the kind of public metadata
> that is allowed.

Reviewer: ```suggestion
is consistent across the entire ecosystem.
```

server issue a token to the client.

~~~

>   Client(pkS, m, info, cMetadata)                        Server(skS, pkS, sMetadata)

Reviewer: Let's spell out these parameters? That is, `cMetadata` => `clientMetadata`.

Reviewer: Also, shouldn't both parties have client and server metadata as input?

Author: Yes! changed everywhere.

opaque PrivateKey<1..2^16-1>
~~~

### Public Metadata {#pp-public-metadata}

We use the following type to describe the metadata that can be added

> by the server and/or client.

Reviewer: ```suggestion
by the server or client.
```

~~~

> opaque Metadata<1..2^16-1>

Reviewer: This means that metadata is length-prefixed, which isn't the case as described in text above. Maybe we just say that it's an arbitrary-length byte string and leave it at that? Then we could delete this section entirely. (See the proposed suggestions below!)

Author: Yeah.. I didn't know how to make it as long as wanted in the TLS language. Changed now ;)

- `pkS`: A server `PublicKey`.
- `skS`: A server `PrivateKey`.
- `req`: An `IssuanceRequest` struct.

> - `sMetadata`: A optional server metadata `Metadata`.
> - `cMetadata`: A optional client metadata `Metadata`.

Reviewer: ```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

- `pkS`: An server `PublicKey`.
- `input`: An `IssuanceInput` struct.
- `resp`: An `IssuanceResponse` struct.

> - `sMetadata`: A optional server metadata `Metadata`.
> - `cMetadata`: A optional client metadata `Metadata`.

Reviewer: ```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

- `info`: An `opaque<1..2^16-1>` type corresponding to data that is
  linked to the redemption. See {{client-info}} for advice on how to
  construct this.

> - `sMetadata`: A optional server metadata `Metadata`.
> - `cMetadata`: A optional client metadata `Metadata`.

Reviewer: ```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```

- `pkS`: An server `PublicKey`.
- `skS`: An server `PrivateKey`.
- `req`: A `RedemptionRequest` struct.

> - `sMetadata`: A optional server metadata `Metadata`.
> - `cMetadata`: A optional client metadata `Metadata`.

Reviewer: ```suggestion
- `sMetadata`: A optional server metadata, an arbitrary-length byte string.
- `cMetadata`: A optional client metadata, an arbitrary-length byte string.
```
